File: binarytree.erl
-module(binarytree).
-compile(export_all).

init() ->
    register(root, spawn(binarytree,server,[])).

server() ->
    server(nil, nil, nil).
server(Left, Value, Right) ->
    receive
        {add, Data} ->
            if
                Value =:= nil ->
                    io:format("Valuenil"),
                    server(Left, [Data], Right);
                Value > Data ->
                    if 
                        Left =:= nil ->
                            io:format("Creating left node"),
                            NewPid = spawn(binarytree, server, [nil, [Data], nil]),
                            server(NewPid, Value, Right);
                        true ->
                            io:format("~p~n",[Left]),
                            Left ! {add, Data},
                            server(Left, Value, Right)
                    end;

                Value < Data ->
                    if
                        Right =:= nil ->
                            io:format("Creating right node"),
                            NewPid = spawn(binarytree, server, [nil, [Data], nil]),
                            server(Left, Value, NewPid);
                        true ->
                            Right ! {add, Data},
                            server(Left, Value, Right)
                    end
            end;
        {val, Who} ->
            Who ! Value,
            server(Left, Value, Right);

        {left, Who} ->
            Who ! {val, Value},
            if 
                Left /= nil ->
                    Left ! {left, Who};
                Left =:= nil ->
                    Who ! fin
            end,
            server(Left, Value, Right);

        {right, Who} ->
            Who ! {val, Value},
            io:format("~p~n",[Right]),
            if
                Right /= nil ->
                    Right ! {right, Who};
                Right =:= nil ->
                    Who ! fin
            end,
            server(Left, Value, Right)            
    end.


walk() ->
    Pid = spawn(binarytree, walk, [[], self()]),
    root ! {left, Pid},
    receive
        {fin, Left} ->
            ok
    end,
    root ! {right, Pid},
    receive
        {fin, Right} ->
            ok
    end,
    Left++Right.
        
walk(Values, Who) ->
    receive
        {val, Value} ->
            io:format("~p~n",[Values]),
            walk(Values ++ Value, Who);
        fin ->
            io:format("Finishing accumulating"),
            Who ! {fin, Values}
    end.

add(Data) ->
    root ! {add, Data}.

File: binarytree.cpp
#include <iostream>
#include <vector>
#include <stdio.h>

struct Tree {
  int Value;
  Tree *Left;
  Tree *Right;
};

class BinaryTree {

public:
  BinaryTree();
  void Add(int);
  std::vector<int> Walk();

private:
  Tree *root;
  void add(int i, Tree *leaf);
  std::vector<int> walk(Tree *leaf);
};

BinaryTree::BinaryTree( ) {
  root = NULL;
}

void BinaryTree::Add(int i) {
  if (root == NULL) {
	root = new Tree;
	root->Value=i;
	return;
  }

  if (root->Value < i) {
	return add(i, root->Right);
  }
  add(i, root->Left);
  return;
}

void BinaryTree::add(int i, Tree *leaf) {
  if (leaf == NULL) {
	leaf = new Tree;
	leaf->Value = i;
	return;
  }
  if (leaf->Value < i) {
	return add(i, leaf->Right);
  }
  return add(i, leaf->Left);
}

std::vector<int> BinaryTree::Walk() {

  std::vector<int> out;
  if (root == NULL) { 
	return out;
  }
  std::vector<int> right;
  std::vector<int>::const_iterator x;
  right = walk(root->Right);
  for (x=right.begin();x!=right.end();++x) {
	out.push_back((*x));
  }
  out.push_back(root->Value);
  std::vector<int> left;
  left = walk(root->Left);
  for (x=left.begin();x!=left.end();++x) {
	out.push_back((*x));
  }
  return out;
}

std::vector<int> BinaryTree::walk(Tree *leaf) {
  std::vector<int> out;
  if (leaf == NULL) { 
	return out;
  }

  std::vector<int> right;
  std::vector<int>::const_iterator x;
  right = walk(leaf->Right);
  for (x=right.begin();x!=right.end();++x) {
	out.push_back((*x));
  }
  out.push_back(leaf->Value);
  std::vector<int> left;
  left = walk(leaf->Left);
  for (x=left.begin();x!=left.end();++x) {
	out.push_back((*x));
  }
  return out;
}

int main() {
  BinaryTree btree;
  btree.Add(5);
  btree.Add(-1);
  btree.Add(3);
  btree.Add(4);
  std::vector<int> tree;
  std::vector<int>::const_iterator x;
  tree = btree.Walk();
  for (x=tree.begin();x!=tree.end();++x) {
	std::cout << (*x) << "\n";
  }
}
